Disabled CTE generation when snippets are used in non-SELECT queries

nooodle-soup opened the pull request:
## Describe your changes
Snippets now when used in anything but a SELECT statement gives a warning and disables CTE generation.

## Issue number

Closes #651 
Closes #652 

## Checklist before requesting a review

- [x] Performed a self-review of my code
- [x] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [x] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing)
- [x] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)



<!-- readthedocs-preview jupysql start -->
----
:books: Documentation preview :books:: https://jupysql--878.org.readthedocs.build/en/878/

<!-- readthedocs-preview jupysql end -->

edublancas commented on src/sql/magic.py:
```diff
@@ -406,22 +408,58 @@ def interactive_execute_wrapper(**kwargs):
                     "Cannot use --with with CTEs, remove --with and re-run the cell"
                 )
 
+            """
```
This code is wrapped in a string (I guess you want to delete it?), please double check your PRs before requesting a review.

nooodle-soup replied:
It is logic related to snippet expansion that I was not sure if I should implement yet. I'll remove it.

edublancas commented on src/sql/magic.py:
```diff
@@ -406,22 +408,58 @@ def interactive_execute_wrapper(**kwargs):
         # only expand CTE if this is not a CTE itself
         if not is_cte:
             if args.with_:
+                query_type = (
+                    sqlparse.parse(command.sql)[0].get_type()
+                    if sqlparse.parse(command.sql)
+                    else None
+                )
```
this same block (or similar one) is repeated in many places, can't we create a function? Also, I don't see any unit tests

nooodle-soup replied:
On it.

nooodle-soup commented:
The checks are failing for some reason I cannot understand. They're files that I haven't touched and the pytest tests run perfectly on my local machine. I am putting it up for review for some inputs @edublancas

edublancas commented:
based on the logs, the image tests are failing (jupysql has some plotting capabilities). you can download the image artifacts from [here](https://github.com/ploomber/jupysql/actions/runs/6194732816?pr=878) (scroll down and you'll see an artifacts section). I downloaded them and noticed that the difference between the expected and the produced image is neglible.

my guess is that the issue is the latest matplotlib release: https://github.com/matplotlib/matplotlib/releases/tag/v3.7.3

to test this, you can try pinning the [matplotlib](https://github.com/ploomber/jupysql/blob/202f0431b313f2a8f0d5ab57ab2cc673a2ece502/setup.py#L47) dependency for a previous release. if that fixes it, please open a new issue so we fix the references images

edublancas commented on src/sql/magic.py:
```diff
@@ -639,6 +669,79 @@ def _persist_dataframe(
+        )
+        return query_type
+
+    def remove_leading_parentheses(self, sql_command):
```
this issue is supposed to close

https://github.com/ploomber/jupysql/issues/651
and
https://github.com/ploomber/jupysql/issues/652

is this parenthesis removing logic relevant to those? I'm a bit confused

nooodle-soup replied:
This is only needed when someone might use a CTE in duckdb in the following manner:

```
(WITH my_penguins as (
    SELECT * from penguins.csv
) SELECT * from my_penguins)
```

Line 406-408 where we check for is_cte would fail in the above case if there is no removing the parentheses.

edublancas replied:
We discussed this parenthesis on a different issue, but I forgot the conclusion (I found this https://github.com/ploomber/jupysql/issues/879 but I think we discussed this somewhere else). In general, we try not to add new requirements to existing issues because it makes it harder to review the code.

also, I feel like we're mixing two separate issues here: CTE generation and SQL queries wrapped in parenthesis. these are two independent problems that we should tackle independently. so please remove this parenthesis handling logic and open an issue so we can discuss what we do about it

nooodle-soup replied:
I've removed it. Opening a new issue regarding the same.

nooodle-soup commented:
I set matplotlib to verison 3.7.2 and the tests passed. I'll open an issue regarding that.

edublancas commented on CHANGELOG.md:
```diff
@@ -2,6 +2,7 @@
 
 ## 0.10.2dev
 
+* [Fix] Fixed snippet behaviour with queries (#651, #652)
```
changelog entries should be more detailed.

example: `Disabled CTE generation when detecting a %%sql cell with an existing CTE `

edublancas commented on src/sql/magic.py:
```diff
@@ -410,18 +414,43 @@ def interactive_execute_wrapper(**kwargs):
+
+                    if query_type != "SELECT":
+                        display.message_warning(
+                            f"Cannot use snippets with {query_type}"
```
the message should be clearer, something like:

```
Your query is using the identifiers 'a', 'b', 'c', which are stored snippets, but snippets only work with SELECT statements, hence, CTE generation is disabled
```

nooodle-soup replied:
I've added better warning messages

edublancas commented on src/sql/magic.py:
```diff
@@ -410,18 +414,43 @@ def interactive_execute_wrapper(**kwargs):
                 )
 
+            if query_type != "SELECT":
+                display.message_warning("Cannot use snippets in CTEs")
```
same here, message should be clearer

nooodle-soup replied:
I've added better warning messages

edublancas commented on src/sql/magic.py:
```diff
@@ -410,18 +414,43 @@ def interactive_execute_wrapper(**kwargs):
+            if query_type != "SELECT":
+                display.message_warning("Cannot use snippets in CTEs")
+
+            dependencies = self._store.infer_dependencies(
+                command.sql_original, args.save
+            )
+
+            parentheses_content = re.findall(
+                r"\((.*)\)", command.sql_original.replace("\n", " ")
+            )
+            for query in parentheses_content:
+                for dependency in dependencies:
+                    if dependency in query:
+                        display.message_warning(
+                            f"Cannot use snippet {dependency} in CTE"
+                        )
```
can you add some comments here? I'm not sure what this chunk of code is for

nooodle-soup replied:
I've added comments

edublancas commented on src/tests/test_magic.py:
```diff
@@ -1925,6 +1925,7 @@ def test_whitespaces_linebreaks_near_first_token(ip, query):
 
     ip.user_global_ns["query"] = query
     out = ip.run_cell("%sql {{query}}").result
+    print(out)
```
please check your PRs before requesting a review, this print statement looks like it was a debugging thing

nooodle-soup replied:
Sorry about that! 😞

edublancas commented on src/tests/test_magic.py:
```diff
@@ -1987,3 +1988,29 @@ def test_accessing_previously_nonexisting_file(ip_empty, tmp_empty, capsys):
+select * from languages where rating < 1"""
+    )
+
+    with pytest.raises(UsageError) as _:
```
looks like this is testing the create table case but you're not testing the case when the user has a CTE. please write more tests to check for any other relevant scenarios.

nooodle-soup replied:
Added more tests

edublancas commented on src/sql/magic.py:
```diff
@@ -639,6 +669,79 @@ def _persist_dataframe(
             table_name=table_name, data_frame=frame, if_exists=if_exists, index=index
         )
 
+    def get_query_type(self, command: SQLCommand, original: bool):
```
there's no need for this function to be an instance method, it can be a separate function. so make it an independent function and add some unit tests

nooodle-soup replied:
Done

edublancas reviewed (changes requested):
when addressing PR comments, don't forget to add links to the relevant changes, so it's easier for the reviewer to see the relevant pieces: https://ploomber-contributing.readthedocs.io/en/latest/contributing/responding-pr-review.html

nooodle-soup commented:
I've added the links to the requested code changes.

edublancas commented on src/sql/magic.py:
```diff
@@ -401,6 +406,9 @@ def interactive_execute_wrapper(**kwargs):
                 "is automatically set as the connection alias"
             )
 
+        # TODO: DuckDB allows queries to be wrapped in parentheses
+        # which needs to be accounted for here. If there is a (WITH ....)
+        # query, then there is an issue of is_CTE not being true.
```
remove this. if we already have an open issue, there's no need for this note

nooodle-soup replied:
Will do!

nooodle-soup replied:
Removed the comments

edublancas commented on src/sql/magic.py:
```diff
@@ -410,18 +418,57 @@ def interactive_execute_wrapper(**kwargs):
                 else:
                     with_ = None
         else:
+            query_type = get_query_type(command.sql_original)
```
you're calling get_query_type two times in two `if` branches, maybe it'd be  better to just call it at the top?

edublancas replied:
actually, this runs when  `is_cte` is True, hence, you're guaranteed to get to this line if you're inside a CTE. do you still need to run `get_query_type`?

nooodle-soup replied:
We do need to check the query type even in the case where is_cte is true as the following will evaluate to a `CREATE` query type and not `SELECT`.

```
%%sql
WITH my_penguins AS (
    SELECT * FROM penguins.csv
) CREATE TABLE my_penguins_2 AS (
    SELECT * FROM my_penguins
)
```

edublancas commented on src/sql/magic.py:
```diff
@@ -410,18 +418,57 @@ def interactive_execute_wrapper(**kwargs):
+            # The following bit of code checks for dependencies within the parentheses
+            # We do not care about the name outside the parentheses being the same as
+            # another table / snippet here and only check if there are dependencies
+            # in the parentheses.
+            # This is for generating more relevant warnings.
+            parentheses_content = re.findall(
+                r"\((.*)\)", command.sql_original.replace("\n", " ")
+            )
+
+            dependency_in_CTE = []
+            for query in parentheses_content:
+                for dependency in dependencies:
+                    if dependency in query:
+                        dependency_in_CTE.append(dependency)
```
please move this to a separate function so we can unit test it.

also, this doesn't seem like the best way to parse this. check how we're inferring dependencies e.g., detect `some_table` in `SELECT * FROM some_table` using sqlglot. that sounds like a better way to parse this

nooodle-soup replied:
I could use that method, but it would become difficult to find out which tables in the query are important to us here.

For example, if we have the following:

```
WITH outer_table AS (
    SELECT * FROM some_snippet
) SELECT * FROM outer_table
```

There are two tables here, `some_snippet` and `outer_table`.
We only want to know if `some_snippet` is a snippet or not. We do not care if `outer_table` is a snippet because we are binding it to the value of `SELECT * FROM some_snippet` in our query, so it has a value in the local scope regardless of any existing snippet in the global scope named `outer_table`. This makes it irrelevant to the error message, which is why I have checked for dependencies within each parentheses only.

nooodle-soup replied:
Made it into a function
[function link](https://github.com/ploomber/jupysql/pull/878/files#diff-527caf90dd08297f5253e826bd98fea0b6f4fbd44fe9f20472976f77868dac06R689-R706)
Added tests
[test link](https://github.com/ploomber/jupysql/pull/878/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R2090-R2114)

edublancas commented on src/sql/magic.py:
```diff
@@ -410,18 +418,57 @@ def interactive_execute_wrapper(**kwargs):
+                        dependency_in_CTE.append(dependency)
+
+            if dependency_in_CTE:
+                if query_type != "SELECT":
+                    display_message = IF_NOT_SELECT_MESSAGE
+                else:
+                    display_message = IF_SELECT_MESSAGE
```
isn't this branch only running when `is_cte` is true? or maybe I'm missing something. under which cases you fall under one `if` condition or the other?

nooodle-soup replied:
One case is when the CTE is used in a CREATE query like follows:

```
%%sql
WITH my_penguins AS (
    SELECT * FROM penguins.csv
) CREATE TABLE my_penguins_2 AS (
    SELECT * FROM my_penguins
)
```

The query type for this will evaluate to `CREATE`.

A regular query like the following:

```
%%sql
my_penguins AS (
    SELECT * FROM penguins.csv
) SELECT * FROM my_penguins
```

will evaluate to query type of `SELECT`.

edublancas commented on src/tests/test_magic.py:
```diff
@@ -1987,3 +1987,100 @@ def test_accessing_previously_nonexisting_file(ip_empty, tmp_empty, capsys):
     assert expected in out
+
+
+def test_warn_when_using_snippets_in_non_SELECT_command(ip_empty, capsys):
```
looks like you're testing many things at once, use pytest's `parametrize` instead

nooodle-soup replied:
[Fixed](https://github.com/ploomber/jupysql/pull/878/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R1992-R2036)

edublancas commented on src/tests/test_magic.py:
```diff
@@ -1987,3 +1987,100 @@ def test_accessing_previously_nonexisting_file(ip_empty, tmp_empty, capsys):
+    )
+
+
+def test_get_query_type():
```
same here, use parametrize

nooodle-soup replied:
[Fixed](https://github.com/ploomber/jupysql/pull/878/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R2058-R2080)

edublancas commented on src/tests/test_magic.py:
```diff
@@ -1987,3 +1987,100 @@ def test_accessing_previously_nonexisting_file(ip_empty, tmp_empty, capsys):
+
+
+def test_get_query_type():
+    assert (
```
also, add one test when calling this with None or invalid SQL, in such case we should return None

nooodle-soup replied:
[Fixed](https://github.com/ploomber/jupysql/pull/878/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R2077-R2084)

nooodle-soup commented:
@edublancas I have made the changes but some tests are failing and I cannot figure out why.
Can you please take a look and let me know where the issue is?

edublancas commented:
I looked at the test:

```
        result = ip_with_dynamic_db.run_cell(
            f"""%%sql
    SELECT * FROM {__TABLE_NAME__};
    """
        ).result
    
>       assert len(result) == 3
E       TypeError: object of type 'NoneType' has no len()

```

`result` is `None`, and `len(None)` breaks. I'd first try investigating if your code introduced the issue or if it's an issue with an external dependency (you can try reproducing it locally or open a dummy PR to trigger the CI and see if tests pass or break with the same error).

nooodle-soup commented:
@edublancas I've fixed the issues
Please have a look and let me know if everything is okay
Thanks!

edublancas commented on src/sql/magic.py:
```diff
@@ -640,6 +674,38 @@ def _persist_dataframe(
+    """
+    Returns the query type of the original sql command
+    """
+    query_type = (
+        sqlparse.parse(command)[0].get_type() if sqlparse.parse(command) else None
```
this is inefficient because it'll run sqlparse.parse twice, better to execute it only once

nooodle-soup replied:
[Fixed](https://github.com/ploomber/jupysql/pull/878/files#diff-527caf90dd08297f5253e826bd98fea0b6f4fbd44fe9f20472976f77868dac06R677-R678)

edublancas commented on src/sql/magic.py:
```diff
@@ -640,6 +674,38 @@ def _persist_dataframe(
+
+def get_dependency_in_CTE(sql_original, dependencies):
+    """
+    The following bit of code checks for dependencies within the parentheses
```
this docstring is pretty long, we can remove things like "the following bit of code checks"

nooodle-soup replied:
I've removed the function entirely and provided a better, more generic warning instead

nooodle-soup replied:
[Fixed](https://github.com/ploomber/jupysql/pull/878/files#diff-527caf90dd08297f5253e826bd98fea0b6f4fbd44fe9f20472976f77868dac06R449-R451)

edublancas commented on src/sql/magic.py:
```diff
@@ -640,6 +674,38 @@ def _persist_dataframe(
+    return query_type
+
+
+def get_dependency_in_CTE(sql_original, dependencies):
```
do not use capital letters in function names

nooodle-soup replied:
This function is no longer in use as I've removed it and provided a more generic warning.

nooodle-soup replied:
[Fixed](https://github.com/ploomber/jupysql/pull/878/commits/a204e262aa93b6bf06494dd2e06c1f27b25e0878)

edublancas commented on src/sql/magic.py:
```diff
@@ -640,6 +674,38 @@ def _persist_dataframe(
+    in the parentheses.
+    This is for generating more relevant warnings.
+    """
+
```
I'm confused about what this function does. based on the code, this is executed when the `is_cte` flag is True, hence, it executes when the user's query begins with `WITH`, and it'll parse anything that's within parenthesis

e.g.,

```sql
with something (
 /* grab everything from here */
)
```

is this right?

if so, I'm confused because of the unit tests has a `CREATE TABLE AS ( ...)` as input, and it doesn't seem like this function will ever be executed with a `CREATE TABLE AS (...)` statement

Also, the unit tests only contain check when the CTE contains a single statement, but this implementation will yield weird results when there is more than one:


```
In [8]: sql_original = """
   ...:
   ...: with something as(
   ...: a
   ...: ), another as(
   ...: b)
   ...: """

In [9]:

In [9]: parentheses_content = re.findall(r"\((.*)\)", sql_original.replace("\n", " "))

In [10]: parentheses_content
Out[10]: [' a ), another as( b']
```

do we want `parenthesis_content` to be equal to `[' a ), another as( b']`? seems like an odd result.

I think at this point, it might be better to just delete this code and show a generic warning, because from what I see, the only purpose of this is to show better warnings.

nooodle-soup replied:
@edublancas I did write this logic for better warning only. I can improve it if you think we need detailed warnings, otherwise I can show a generic one.

nooodle-soup replied:
@edublancas I've removed the logic and provided a more generic warning instead

nooodle-soup replied:
[Fixed](https://github.com/ploomber/jupysql/pull/878/commits/a204e262aa93b6bf06494dd2e06c1f27b25e0878)

nooodle-soup commented on setup.py:
```diff
@@ -18,7 +18,7 @@
     "prettytable",
     # IPython dropped support for Python 3.8
-    "ipython<=8.12.0; python_version <= '3.8'",
+    "ipython; python_version <= '3.8'",
```
This change stopped the breaking issues for some tests
I'll investigate more and give an update.

nooodle-soup replied:
The errors disappeared with the latest update of IPython

nooodle-soup replied:
@edublancas  I don't have the link for the commit where I fixed this as I had to reset the git repo in my fork and pulled from master again to create a new branch that did not have this change.

edublancas commented:
@nooodle-soup what's the status here?

nooodle-soup commented:
@edublancas I'm waiting for a review
I've fixed all the issues that we discussed. I removed the get_dependencies_in_CTE and instead provided a more generic warning message.

nooodle-soup commented on setup.py:
```diff
@@ -37,7 +37,7 @@
     "pyarrow",
     "invoke",
-    "pkgmt",
+    "pkgmt==0.7.3",
```
I've removed it as the latest IPython update fixed it

edublancas commented:
@nooodle-soup remember to follow the guidelines and respond each PR observation with a link: https://ploomber-contributing.readthedocs.io/en/latest/contributing/responding-pr-review.html

also, you need to explicitly request a review, otherwise I don't know if the PR is ready for another review or not. please request the review either by using the github button or by adding a comment and tagging me

nooodle-soup commented:
> @nooodle-soup remember to follow the guidelines and respond each PR observation with a link: https://ploomber-contributing.readthedocs.io/en/latest/contributing/responding-pr-review.html

@edublancas I thought I added links to almost all the responses where code was modified. Should I be adding links to the removed code as well? I assumed the code no longer existed so I should not be adding a link to that change 😓 . I'll add links there as well if you say so.

I'll ask for reviews by tagging from now on! 😄

edublancas commented:
if you're removing code. you can just add a link to the commit that deleted it. I do that (one commit per PR observation and then respond to the PR reviewer with links to each commit)

nooodle-soup commented:
> if you're removing code. you can just add a link to the commit that deleted it. I do that (one commit per PR observation and then respond to the PR reviewer with links to each commit)

Okay. I'll do the same from now on! 👍🏽

nooodle-soup commented:
@edublancas  I've added links to all the changes I could. Please have a look when you can and let me know if I can make my review better! Thanks!

edublancas commented on src/tests/test_magic.py:
```diff
@@ -1987,3 +1987,102 @@ def test_accessing_previously_nonexisting_file(ip_empty, tmp_empty, capsys):
+SELECT * FROM my_penguins",
+            "SELECT",
+        ),
+        (
+            " WITH my_penguins ( \
+    SELECT * FROM penguins.csv \
+) \
+SELECT * FROM my_penguins",
+            None,
+        ),
```
this example is pretty hard to read, took me sometime to realize it's missing the `AS`. if you're testing invalid SQL, add some obviously invalid SQL code so the reader doesn't have to guess what you're testing

nooodle-soup replied:
[Fixed](https://github.com/ploomber/jupysql/pull/878/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R2097-R2100)

edublancas replied:
is this link right? it's taking me to all the changed files, not to a specific set of lines or file

nooodle-soup replied:
[Fixed](https://github.com/ploomber/jupysql/pull/878/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R2082-R2090)
Updated link @edublancas

edublancas commented on setup.py:
```diff
@@ -19,7 +19,7 @@
     # IPython dropped support for Python 3.8
     "ipython<=8.12.0; python_version <= '3.8'",
-    "ipython<8.16.0",
+    "ipython",
```
remove this, we already have ipython as dependency (line 21), we don't need this one

nooodle-soup replied:
[Fixed](https://github.com/ploomber/jupysql/pull/878/commits/3cce632b8d4c8f162b3a09b87c6fade4db0d15e4)

edublancas replied:
the link seems broken

![image](https://github.com/ploomber/jupysql/assets/989250/c11f2445-0d06-4e66-b602-bc8a11396245)

nooodle-soup replied:
[Fixed](https://github.com/ploomber/jupysql/pull/878/commits/d6002ac0e223e15ca368c27bf22b6bf70d821158)

Updated the link

edublancas commented on src/tests/test_magic.py:
```diff
@@ -1987,3 +1987,102 @@ def test_accessing_previously_nonexisting_file(ip_empty, tmp_empty, capsys):
+    "query, query_type",
+    [
+        (
+            "CREATE TABLE penguins AS ( \
+    WITH my_penguins AS ( \
```
do not use `\` for splitting strings as it makes the tests hard to read, better to define multiline strings with `"""`

nooodle-soup replied:
[Fixed](https://github.com/ploomber/jupysql/pull/878/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R2072-R2107)

edublancas replied:
please change it to `"""` (triple quotes)

nooodle-soup replied:
[Fixed](https://github.com/ploomber/jupysql/pull/878/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R2059-R2092)

edublancas commented:
@neelasha23 please run some manual tests and let us know if everything is ok

nooodle-soup commented:
@edublancas 
I've implemented the changes and the PR is ready for review
Thanks!

neelasha23 commented on setup.py:
```diff
@@ -19,7 +19,6 @@
     "prettytable",
     # IPython dropped support for Python 3.8
     "ipython<=8.12.0; python_version <= '3.8'",
-    "ipython",
```
why is ipython removed?

nooodle-soup replied:
@edublancas suggested the removal
Is there any issue you faced in your manual tests?

neelasha23 commented:
1. This still seems to throw an error and it's not very clear what should be the user's next step:
<img width="1064" alt="Screenshot 2023-10-10 at 12 13 30 PM" src="https://github.com/ploomber/jupysql/assets/29096135/6e213e7b-f5e7-4a0b-8e11-a6a63f6de460">

Specifying `--with` will also throw error:

<img width="929" alt="Screenshot 2023-10-10 at 12 21 09 PM" src="https://github.com/ploomber/jupysql/assets/29096135/d721d427-cc65-4f94-9950-298f120e9026">

2. It displays warning that CTE generation is disabled but the query is displaying some results. How is this working?
<img width="1061" alt="Screenshot 2023-10-10 at 12 31 27 PM" src="https://github.com/ploomber/jupysql/assets/29096135/84e734ed-b44c-4877-96c5-9c9fbd2cda7d">

3. Here also it's not clear what the user needs to do as the next step
<img width="1051" alt="Screenshot 2023-10-10 at 12 33 04 PM" src="https://github.com/ploomber/jupysql/assets/29096135/9420ab5c-15a8-40a3-baaf-78a99ec18357">

nooodle-soup commented:
@neelasha23

If you saved a snippet called `my_penguins`, and also used it in a CTE in some way like 
```
WITH my_penguins AS (
    SELECT * FROM penguins.csv
) SELECT * FROM my_penguins
```

We want the code to run here because it is valid SQL that does not depend on any external snippet. Even though the name is the same, the `my_penguins` snippet is not to be used here and instead `my_snippet` temporary table which refers to `SELECT * FROM penguins.csv` is to be used.

neelasha23 commented:
> @neelasha23
> 
> If you saved a snippet called `my_penguins`, and also used it in a CTE in some way like
> 
> ```
> WITH my_penguins AS (
>     SELECT * FROM penguins.csv
> ) SELECT * FROM my_penguins
> ```
> 
> We want the code to run here because it is valid SQL that does not depend on any external snippet. Even though the name is the same, the `my_penguins` snippet is not to be used here and instead `my_snippet` temporary table which refers to `SELECT * FROM penguins.csv` is to be used.

So maybe we should display warning something like:  Running query with temporary_table `my_snippet` instead of saved snippet `my_snippet` . wdyt @edublancas

nooodle-soup commented:
@neelasha23 I've updated my previous point
